MapboxCoreNavigation/RouteController.swift

```diff
@@ -333,6 +342,50 @@ extension RouteController: CLLocationManagerDelegate {
         }
     }
     
+    func checkForFasterRoute(from location: CLLocation) {
+        guard let currentUpcomingManeuver = routeProgress.currentLegProgress.upComingStep else { return }
+        
+        guard let lastLocationDate = lastLocationDate else {
+            self.lastLocationDate = location.timestamp
+            return
+        }
+
+        // Only check ever 2 minutes for faster route
+        guard location.timestamp.timeIntervalSince(lastLocationDate) >= 15 else { return }
+        let durationRemaining = routeProgress.durationRemaining
+        let currentAlertLevel = routeProgress.currentLegProgress.alertUserLevel
+        
+        func rerouteIfFaster(onto route: Route) {
+            // Only use new route if it's at least 10% faster
+            if route.expectedTravelTime <= 0.9 * durationRemaining {
+                // If the upcoming maneuver in the new route is the same as the current upcoming maneuver, don't announce it
+                if let firstLeg = route.legs.first, firstLeg.steps.count > 2 {
+                    routeProgress = RouteProgress(route: route, legIndex: 0, alertLevel: currentUpcomingManeuver == firstLeg.steps[1] ? currentAlertLevel : .none)
+                    delegate?.routeController?(self, didRerouteAlong: route)
+                }
+            }
+        }
+        
+        getDirections(from: location) { [weak self] (route, error) in
+            guard let strongSelf = self else { return }
+            guard let route = route else { return }
+            guard let firstLeg = route.legs.first, firstLeg.steps.count > 2 else { return }
+            strongSelf.lastLocationDate = nil
+            
+            if firstLeg.steps[1].expectedTravelTime <= 70 {
+                let course = currentUpcomingManeuver.finalHeading ?? currentUpcomingManeuver.initialHeading ?? location.course
+                let maneuverLocation = CLLocation(coordinate: currentUpcomingManeuver.maneuverLocation, altitude: 1, horizontalAccuracy: 1, verticalAccuracy: 1, course: course, speed: 1, timestamp: Date())
+                
+                strongSelf.getDirections(from: maneuverLocation) { (route, error) in
```

Comment:
That’s entirely up to the logic that follows for determining whether the new route is distinct from the existing route. Currently it doesn’t even check whether the two routes are distinct; it just goes with the new one as long as it’s faster. Conceptually, we should only use the new route if doing so would keep the contents of the turn banner the same for now.

Moreover, in terms of the alert level, it compares the `currentUpcomingManeuver`’s coordinates with those of `firstLeg.steps[1]`, which I think is incorrect. Consider the following scenario:

* `currentManeuver` has a few miles left.
* `currentUpcomingManeuver` is a right turn at Albuquerque.
* `firstLeg.steps[1]` is a departure at the user’s current location.
* `firstLeg.steps[2]` is a right turn at Albuquerque.

As I pointed out in https://github.com/mapbox/mapbox-navigation-ios/pull/346#discussion_r127333181, we should be comparing the maneuver type, direction, and coordinates of `currentUpcomingManeuver` and `firstLeg.steps[2]`. Otherwise, we are indeed guaranteed a reroute.